source/common/router/router.cc

```diff
@@ -485,7 +486,21 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Fetch a connection pool for the upstream cluster.
-  Http::ConnectionPool::Instance* conn_pool = getConnPool();
+  Http::ConnectionPool::Instance* conn_pool;
+  auto url_str = headers.Host()->value().getStringView();
+  const auto parsed_authority = Http::Utility::parseAuthority(url_str.data());
+  const auto http_protocol_options = cluster_->httpProtocolOptions();
+
+  if (http_protocol_options != absl::nullopt && http_protocol_options->auto_sni() &&
```

Comment:
perf nit: refactor to avoid all of the authority parsing work unless you are actually using auto sni?